### PR TITLE
update resnet of faster-rcnn configuration

### PR DESCRIPTION
**Pull request by fullfanta**
I updated faster-rcnn's resnet configuration.

1. config.py
in config.py, FIXED_PARAMS are defined and they are searched in module.py as (https://github.com/dmlc/mxnet/blob/master/example/rcnn/rcnn/core/module.py#L43-L49)
```
        if fixed_param_prefix is not None:
            for name in self._symbol.list_arguments():
                for prefix in self._fixed_param_prefix:
                    if name.startswith(prefix):
                        fixed_param_names.append(name)
```
But, there are no params start with 'gamma' and 'beta', but 'bn0', 'bn_data'.
So, I updated configuration for resnet.

2. symbol_resnet.py
When I loaded pretrained resnet model and finetune for faster-rcnn, bn_data's gamma becomes not pretrained value but 1 due to fix_gamma=True.

According to the config.py, to use pretrained model's parameters again in (conv0, stage1, bn0, bn_data), fix_gamma should be False.

3. Experiment
I trained resnet101 using end2end method, VOC 07 data, with different configuration.

With previous configuration, I got mAP as 74.33%.
With updated configuration, I got mAP as 75.26%

In detail, with previous configuration,
```
AP for aeroplane = 0.7521
AP for bicycle = 0.7964
AP for bird = 0.7490
AP for boat = 0.6181
AP for bottle = 0.5711
AP for bus = 0.7941
AP for car = 0.8394
AP for cat = 0.8789
AP for chair = 0.5871
AP for cow = 0.8187
AP for diningtable = 0.7141
AP for dog = 0.8578
AP for horse = 0.8587
AP for motorbike = 0.7777
AP for person = 0.7865
AP for pottedplant = 0.4594
AP for sheep = 0.7270
AP for sofa = 0.7529
AP for train = 0.7722
AP for tvmonitor = 0.7540
Mean AP = 0.7433
```

with updated configuration, 
```
AP for aeroplane = 0.7669
AP for bicycle = 0.8166
AP for bird = 0.7678
AP for boat = 0.6317
AP for bottle = 0.5984
AP for bus = 0.7827
AP for car = 0.8442
AP for cat = 0.8623
AP for chair = 0.5904
AP for cow = 0.8179
AP for diningtable = 0.7044
AP for dog = 0.8682
AP for horse = 0.8445
AP for motorbike = 0.7930
AP for person = 0.7890
AP for pottedplant = 0.4745
AP for sheep = 0.7610
AP for sofa = 0.7598
AP for train = 0.8099
AP for tvmonitor = 0.7685
Mean AP = 0.7526
```

According to the results, I suggest you to update.

### CODE REVIEW

**Comment by piiswrong** [comment]
@precedenceguo

**Comment by precedenceguo** [comment]
Thank you for your contribution.

On the first point, thank you for your notice about this. It was introduced when we process nnvm 0.9.3 changes as in commit https://github.com/precedenceguo/mx-rcnn/commit/410f089ddec6abd6afa453c869e81c36020b8c64
As a consequence, it is possible that resnet results cannot be fully reproduced to my experiments on v0.9.1.
Would you please consider change it back to "prefix in name" instead of changing the config for "name.startswith(prefix)"?

On the second point, I am afraid that "fix_gamma=True" is set for "bn_data" because of the pretrained ResNet-v2 model is trained this way c.f. https://github.com/tornadomeet/ResNet/blob/master/symbol_resnet.py

On the result of improvement, I think by changing to "prefix in name", better performance is possible.

**Comment by piiswrong** [comment]
@precedenceguo Could you submit a PR for the change?

**Comment by fullfanta** [comment]
@precedenceguo , @piiswrong 
I'll test with different parameters such as 1) "prefix in name" vs. "name.startswith(prefix)", 2) fixed_gamma=True vs. False on 'bn_data' and let you know the result as soon as possible.

**Comment by piiswrong** [comment]
any updates?

**Comment by fullfanta** [comment]
@piiswrong , @precedenceguo 

Now, I finished test.

1. I tested in voc07
with fix_gamma=True at 'bn_data',

|    | startswith or in | fixed_params | mAP |  |
| - | --------------  | ------------- | --- | -- |
| 1-1 | startswith          | gamma, beta  | 74.33%| current setting |
| 1-2 | startswith        | bn0, bn_data | 74.85%|
| 1-3 | in                      | gamma, beta | 75.01% |

with fix_gamma=False at 'bn_data',

| | startswith or in | fixed_params | mAP |  |
| - | --------------  | ------------- | --- | -- |
| 1-4 | startswith        | bn0, bn_data | 75.26%|
| 1-5 | in                      | gamma, beta | 75.03% |

in voc07, 1-4 and 1-5 shows best performance.

2. For more test, I used voc07+12.

| | startswith or in | fixed_params | fixed_gamma=True or False | mAP |
| - | --------------- | ------------- | ---------------------------- | ---- |
| 2-1 | startswith | bn0, bn_data | True | 79.48% |
| 2-2 | startswith | bn0, bn_data | False | 80.12% |
| 2-3 | in | gamma, beta | True | 80.14% |
| 2-4 | in | gamma, beta | False | 79.88% |

in voc07+12, 2-3 and 2-4 shows best performance.

As @precedenceguo recommends, changing from 'startswith' to 'in' is a good choice simply and get good results as 75.01% in voc07, 80.14% in voc07+12.

But what I concern is that 1) changing from 'startswith' to 'in' with 'gamma, beta' means all parameters of batch normalization are fixed as pretrained values and are not trained anymore even though it was pretrained using imagenet.

Anyway, I think @precedenceguo's recommendation is good option.
If both of you are agree, I'll make update.

**Comment by piiswrong** [comment]
looks good. Please update

**Comment by piiswrong** [comment]
I think fix gamma should still be true since the weights are originally trained with fix_gamma=True

**Comment by fullfanta** [comment]
@piiswrong , @precedenceguo 
I updated. Please review. :)

File: example/rcnn/rcnn/config.py [hunk 1]
```diff
@@ -134,7 +134,7 @@
 network.resnet.RPN_FEAT_STRIDE = 16
 network.resnet.RCNN_FEAT_STRIDE = 16
 network.resnet.FIXED_PARAMS = ['conv0', 'stage1', 'gamma', 'beta']
-network.resnet.FIXED_PARAMS_SHARED = ['conv0', 'stage1', 'stage2', 'stage3', 'gamma', 'beta']
+network.resnet.FIXED_PARAMS_SHARED = ['conv0', 'stage1', 'stage2', 'stage3', 'bn0', 'bn_data']
```
Comment (piiswrong) [inline example/rcnn/rcnn/config.py]: why change this? Which experiment does this correspond to?

Comment (fullfanta) [inline example/rcnn/rcnn/config.py]: Oh. Sorry. I tried alternative training which is not mentioned here.
It should be 'gamma' and 'beta'. I'll fix it.

**Comment by fullfanta** [comment]
@piiswrong 
I updated again.

**Comment by fullfanta** [comment]
@piiswrong 
Could you please merge if you don't have any problem after review?

**Comment by precedenceguo** [comment]
Detection works I've seen all fix bn gamma and beta so there should be no worry.
`fix_gamma` for `bn_data` is really not important if all bn params are fixed so it could be kept as the way it is trained.
Thanks again!
